### PR TITLE
postgresqlAdvanceThreadProtection

### PR DESCRIPTION
**Pull request by fatima99s**

### CODE REVIEW

File: plugins/azure/postgresqlserver/flexibleServerATP.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+var async = require('async');
+var helpers = require('../../../helpers/azure');
+
+module.exports = {
+    title: 'PostgreSQL Flexible Server Advanced Threat Protection',
+    category: 'PostgreSQL Server',
+    domain: 'Databases',
+    description: 'Ensures Defender is enabled for Advanced Threat Protection on PostgreSQL flexible servers.',
+    more_info: 'Enabling Defender for Cloud provides security alerts on anomalous activities, allowing you to detect potential threats and respond to them as they occur.',
+    recommended_action: 'Ensure Defender is enabled for PostgreSQL Flexible Server',
```
[the remark this answers is not in the document]
Comment (mehakseedat63) [reply plugins/azure/postgresqlserver/flexibleServerATP.js:10]: Screenshot 2024-02-01 at 3 20 36 PM
I don't see ATP for flexible servers, can only see defender which is enabled at subscription level and falls under "open source relational databases" category, please verify this.

**Review by mehakseedat63 — CHANGES_REQUESTED** [review]

File: plugins/azure/postgresqlserver/flexibleServerATP.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (fatima99s) [reply plugins/azure/postgresqlserver/flexibleServerATP.js:10]: we can enable the defender for postgres flexible server at resource level unlike other services if the defender is enable its enabled for all resources of that service , for postgres we can enable defender for one server and disable for others.

File: plugins/azure/postgresqlserver/flexibleServerATP.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
+var async = require('async');
+var helpers = require('../../../helpers/azure');
+
+module.exports = {
+    title: 'PostgreSQL Flexible Server Advanced Threat Protection',
+    category: 'PostgreSQL Server',
+    domain: 'Databases',
+    description: 'Ensures Defender is enabled for Advanced Threat Protection on PostgreSQL flexible servers.',
```
Comment (mehakseedat63) [inline plugins/azure/postgresqlserver/flexibleServerATP.js:8]:
```suggestion
    description: 'Ensures Advanced Threat Protection is enabled on PostgreSQL flexible servers.',
```

File: plugins/azure/postgresqlserver/flexibleServerATP.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    more_info: 'Enabling Defender for Cloud provides security alerts on anomalous activities, allowing you to detect potential threats and respond to them as they occur.',
+    recommended_action: 'Ensure Defender is enabled for PostgreSQL Flexible Server.',
```
Comment (mehakseedat63) [inline plugins/azure/postgresqlserver/flexibleServerATP.js:10]: change the wording for Advanced Threat protection and not defender

**Review by mehakseedat63 — CHANGES_REQUESTED** [review]

File: plugins/azure/postgresqlserver/flexibleServerATP.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                } else {
+                    if (!advancedThreatProtectionSettings.data.length) {
+                        helpers.addResult(results, 2, 'No Advanced Threat Protection setting found', location, server.id);
+                    } else {
+                        let atpEnabled = advancedThreatProtectionSettings.data.find(threadProtectionSetting => 
+                            threadProtectionSetting.state &&
+                            threadProtectionSetting.state.toLowerCase() == 'enabled');
+                        if (atpEnabled) {
+                            helpers.addResult(results, 0,
+                                'Advanced Threat Protection for the PostgreSQL flexible server is enabled', location, server.id);
```
Comment (mehakseedat63) [inline plugins/azure/postgresqlserver/flexibleServerATP.js:54]:
```suggestion
                                'Advanced Threat Protection is enabled for PostgreSQL flexible server', location, server.id);
```

File: plugins/azure/postgresqlserver/flexibleServerATP.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                        } else {
+                            helpers.addResult(results, 2,
+                                'Advanced Threat Protection for the PostgreSQL flexible server is disabled', location, server.id);
```
Comment (mehakseedat63) [inline plugins/azure/postgresqlserver/flexibleServerATP.js:57]:
```suggestion
                                'Advanced Threat Protection is disabled for PostgreSQL flexible server', location, server.id);
```

File: plugins/azure/postgresqlserver/flexibleServerATP.spec.js [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            );
+
+            enableATP.run(cache, {}, callback);
+        });
+
+        it('should give failing result if Advanced Threat Protection for the PostgreSQL flexible server is disabled', function(done) {
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].message).to.include('Advanced Threat Protection for the PostgreSQL flexible server is disabled');
```
Comment (mehakseedat63) [inline plugins/azure/postgresqlserver/flexibleServerATP.spec.js:90]:
```suggestion
                expect(results[0].message).to.include('Advanced Threat Protection is disabled for PostgreSQL flexible server');
```

File: plugins/azure/postgresqlserver/flexibleServerATP.spec.js [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+            );
+
+            enableATP.run(cache, {}, callback);
+        });
+
+        it('should give passing result if Advanced Threat Protection for the PostgreSQL flexible server is enabled', function(done) {
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].message).to.include('Advanced Threat Protection for the PostgreSQL flexible server is enabled');
```
Comment (mehakseedat63) [inline plugins/azure/postgresqlserver/flexibleServerATP.spec.js:107]:
```suggestion
                expect(results[0].message).to.include('Advanced Threat Protection is enabled for PostgreSQL flexible server');
```

**Review by mehakseedat63 — CHANGES_REQUESTED** [review]

**Review by mehakseedat63 — APPROVED** [review]

**Pull request merged**